common/src/main/kotlin/org/imdc/extensions/common/DatasetExtensions.kt

```diff
@@ -293,50 +339,66 @@ object DatasetExtensions {
                 val row = sheet.getRow(i)
 
                 val rowValues = Array<Any?>(columnCount) { j ->
-                    val cell = row.getCell(j + firstColumn)
-
-                    when (cell?.cellType.takeUnless { it == FORMULA } ?: cell.cachedFormulaResultType) {
-                        NUMERIC -> {
-                            if (DateUtil.isCellDateFormatted(cell)) {
-                                if (!typesSet) {
-                                    columnTypes.add(Date::class.java)
-                                }
-                                cell.dateCellValue
-                            } else {
-                                val numericCellValue = cell.numericCellValue
-                                if (BigDecimal(numericCellValue).scale() == 0) {
+                    val cell: Cell? = row.getCell(j + firstColumn)
+                    val typeOverride = typeOverrides[j]
+                    if (typeOverride != null) {
+                        if (!typesSet) {
+                            columnTypes.add(typeOverride)
+                        }
+                        when {
+                            cell == null -> null
+                            typeOverride.isAssignableFrom<Date>() -> cell.dateCellValue
+                            typeOverride.isAssignableFrom<Boolean>() -> cell.booleanCellValue
+                            typeOverride.isAssignableFrom<String>() -> cell.stringCellValue
+                            typeOverride.isAssignableFrom<Int>() -> cell.numericCellValue.toInt()
+                            typeOverride.isAssignableFrom<Long>() -> cell.numericCellValue.toLong()
+                            typeOverride.isAssignableFrom<Number>() -> cell.numericCellValue
+                            else -> throw Py.TypeError("Unable to retrieve R${cell.row.rowNum + 1}C${cell.columnIndex + 1} as a ${typeOverride.simpleName}")
+                        }
```

Comment:
Slightly awkward here, but should still cover most cases.

Reply:
Yes, I am confused by the logic of this line. 
`cell == null -> null`
instead of
`cell != null -> cell`

Wouldn't this only run when the cell is `null`? If the cell is `null,` wouldn't all of the `cell.type` error out?

Reply:
Thanks so much for the explanation! That makes a lot of sense.